python.d/ntp.chart.py

```diff
@@ -0,0 +1,381 @@
+        'lines': [
+            ['rootdisp', 'dispersion', 'absolute', 1, PRECISION]
+        ]},
+    'sys_stratum': {
+        'options': [None, "Stratum (1-15)", "1", 'system', 'ntp.sys_stratum', 'line'],
+        'lines': [
+            ['stratum', 'stratum', 'absolute', 1, PRECISION]
+        ]},
+    'sys_tc': {
+        'options': [None, "Time constant and poll exponent (3-17)", "log2 s", 'system', 'ntp.sys_tc', 'line'],
+        'lines': [
+            ['tc', 'current', 'absolute', 1, PRECISION],
+            ['mintc', 'minimum', 'absolute', 1, PRECISION]
+        ]},
+    'sys_precision': {
+        'options': [None, "Precision", "log2 s", 'system', 'ntp.sys_precision', 'line'],
+        'lines': [
+            ['precision', 'precision', 'absolute', 1, PRECISION]
+        ]}
+}
+
+# Dynamic charts templates
+PEER_PREFIX = 'peer'
+
+PEER_DIMENSIONS = [
+    ['offset', 'Filter offset', 'ms'],
+    ['delay', 'Filter delay', 'ms'],
+    ['dispersion', 'Filter dispersion', 'ms'],
+    ['jitter', 'Filter jitter', 'ms'],
+    ['xleave', 'Interleave delay', 'ms'],
+    ['rootdelay', 'Total roundtrip delay to the primary reference clock', 'ms'],
+    ['rootdisp', 'Total root dispersion to the primary reference clock', 'ms'],
+    ['stratum', 'Stratum (1-15)', '1'],
+    ['hmode', 'Host mode (1-6)', '1'],
+    ['pmode', 'Peer mode (1-5)', '1'],
+    ['hpoll', 'Host poll exponent', 'log2 s'],
+    ['ppoll', 'Peer poll exponent', 'log2 s'],
+    ['precision', 'Precision', 'log2 s']
+]
+
+
+class Peer(object):
+    """
+    Class to hold peer data required in _get_data
+    """
+    def __init__(self, peer_id, name, request):
+        self.id = peer_id
+        self.name = name
+        self.request = request
+
+
+class Service(SocketService):
+    def __init__(self, configuration=None, name=None):
+        SocketService.__init__(self, configuration=configuration, name=name)
+        self.port = 'ntp'
+        self.dgram_socket = True
+        self.request_systemvars = None
+        self.peers = None
+        self.peer_error = 0
+        self.regex_srcadr = re.compile(r'srcadr=([A-Za-z0-9.-]+)')
+        self.regex_data = re.compile(r'([a-z_]+)=([0-9-]+(?:\.[0-9]+)?)(?=,)')
+        self.order = None
+        self.definitions = None
+        self.peer_names = self.configuration.get('peer_names', True)
+        peer_filter_start = r'^((0\.0\.0\.0)|('
+        peer_filter_end = r'))$'
+        peer_filter_default = r'127\..*'
+        peer_filter_custom = str(self.configuration.get('peer_filter', peer_filter_default))
+
+        try:
+            self.regex_peer_filter = re.compile(peer_filter_start + peer_filter_custom + peer_filter_end)
+        except re.error as error:
+            self.error('Pattern compile error: %s, Using defaults.' % str(error))
+            self.regex_peer_filter = re.compile(peer_filter_start + peer_filter_default + peer_filter_end)
```

Comment:
actually i think we should not use defaults in case of pattern compile error. It should warn user about error and exit (return None). What do you think?
if you agree with me move `peer_filter_` stuff to the `check()`

```python
def check():
    peer_filter_custom = self.configuration.get('peer_filter', r'127\..*')
    try:
        self.regex_peer_filter = re.compile(r'^((0\.0\.0\.0)|({0}))$'.format(peer_filter_custom))
    except re.error as error:
        self.error('Pattern compile error: {0}'.format(error))
        return None
```

Reply:
Yes, this is better, I'll change it.